### PR TITLE
Add tensor-based annotation storage to reduce DDP RAM usage with large datasets

### PR DESCRIPTION
**Pull request by spsancti**
Source of the bug: https://github.com/pytorch/pytorch/issues/13246
Solution based on: https://github.com/facebookresearch/detectron2/commit/0cd0e7253f64fc40e9c2ee4617df6e4760d07974

### CODE REVIEW

File: src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py [hunk 1]
```diff
@@ -124,7 +147,9 @@ def _load_annotation(self, sample_id: int) -> dict:
         :return img_path:               Path to the associated image
         """
 
-        annotation = self._annotations[sample_id]
+        start_addr = 0 if sample_id == 0 else self._addr[sample_id - 1].item()
+        end_addr = self._addr[sample_id].item()
+        annotation = pickle.loads(self._annotations[start_addr:end_addr].numpy().data)
```
Comment (NatanBagrov) [inline src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py:152]: We saw slowdown due to pickle load, do we want to make the serialize-parse fix optional?
I mean, eventually a memory leak is  a memory leak, but for small datasets you get overhead whereas without the fix "you'd be fine". @BloodAxe , thoughts?

File: src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py [hunk 2]
```diff
@@ -284,9 +318,9 @@ def parse_coco_into_detection_annotations(
             image_path=image_path,
             image_width=image_width,
             image_height=image_height,
-            ann_boxes_xyxy=ann_box_xyxy[mask],
-            ann_is_crowd=ann_iscrowd[mask],
-            ann_labels=ann_category_id[mask],
+            ann_boxes_xyxy=np.asarray(img_id2ann_box_xyxy[img_id]),
+            ann_is_crowd=np.asarray(img_id2ann_iscrowd[img_id]),
+            ann_labels=np.asarray(img_id2ann_category_id[img_id]),
```
Comment (NatanBagrov) [inline src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py:323]: Already in a different PR, https://github.com/Deci-AI/super-gradients/pull/1888

**Review by NatanBagrov — COMMENTED** [review]
Thanks! See comments inline.

File: src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py [hunk 3]
```diff
@@ -104,8 +114,21 @@ def _setup_data_source(self) -> int:
 
         self.original_classes = list(all_class_names)
         self.classes = copy.deepcopy(self.original_classes)
-        self._annotations = annotations
-        return len(annotations)
+
+        self._annotations = [_serialize(x) for x in annotations]
+
+        del annotations, all_class_names
```
Comment (shaydeci) [inline src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py:120]: Why do we neet to delete all_class_names?

File: src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py [hunk 4]
```diff
@@ -1,23 +1,33 @@
...
 from super_gradients.training.datasets.data_formats.bbox_formats.xywh import xywh_to_xyxy_inplace
-from super_gradients.training.datasets.detection_datasets.detection_dataset import DetectionDataset
 from super_gradients.training.datasets.data_formats.default_formats import XYXY_LABEL
+from super_gradients.training.datasets.detection_datasets.detection_dataset import DetectionDataset
 from super_gradients.training.utils.detection_utils import change_bbox_bounds_for_image_size
 
 logger = get_logger(__name__)
 
 
+def _serialize(data):
```
Comment (shaydeci) [inline src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py:26]: please move this to a more appropriate place.

File: src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py [hunk 5]
```diff
@@ -104,8 +114,21 @@ def _setup_data_source(self) -> int:
...
+
+        self._annotations = [_serialize(x) for x in annotations]
+
+        del annotations, all_class_names
+        gc.collect()
+
+        self._addr = torch.tensor([len(x) for x in self._annotations], dtype=torch.int64)
+        self._addr = torch.cumsum(self._addr, dim=0)
+        self._annotations = torch.concatenate(self._annotations)
+        print("Serialized dataset takes {:.2f} MiB".format(len(self._annotations) / 1024**2))
```
Comment (shaydeci) [inline src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py:126]: Please remove print

File: src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py (hunk 1 above)
Comment (shaydeci) [reply src/super_gradients/training/datasets/detection_datasets/coco_format_detection.py:152]: @NatanBagrov yes IMO

**Review by shaydeci — CHANGES_REQUESTED** [review]
To me this looks quite ready.
I do have one more reqyest: please add a unit test that utilizes this feature so that we see nothing crashes. You can use our data in tests/data/coco2017 (see how we use it for example in tests/unit_tests/preprocessing_unit_test.py).
It can be a simple test that just iterates throught the dataset with use_tensor_backed_storage set.
